### PR TITLE
Serialization Proto for PyTorch

### PR DESCRIPTION
**Pull request by houseroad**
We introduce the serialization format for PyTorch, based on existing ONNX format. (Not reinventing another proto format.)

1) Additional Graphs (called method) are added into ModelProto to support submodule and methods.
2) TensorProto are extended to support external file and strides, offset can be supported by segment.
3) All the weights will be stored in the main graph, with full name "(sub)module_name.weight_name"
4) MetaNetDef are merged into ModelProto.
5) Add Sequence (Tuple, List) and Map types into ValueInfoProto, and clarify how to use TensorProto to represent scalar.

### CODE REVIEW

File: caffe2/proto/torch.proto [hunk 1]
```diff
@@ -416,11 +466,36 @@ message TypeProto {
...
+  // serializer and deserialzier.
+  message SpecialBlob {
+    optional string type_name = 1;
+  }
 
   oneof value {
     // The type of a tensor.
     Tensor tensor_type = 1;
-
+    Squence sequence_type = 4;
```
Comment (li-roy) [inline caffe2/proto/torch.proto]: /Squence/Sequence/

File: caffe2/proto/torch.proto [hunk 2]
```diff
@@ -416,11 +466,36 @@ message TypeProto {
     optional TensorShapeProto shape = 2;
```
Comment (li-roy) [inline caffe2/proto/torch.proto:479]: We're currently storing strides in Dimension of TensorShapeProto. For example, if a tensor is of shape (2, 2) and strides (1, 2), we store 2, 2, 1, 2 in Dimension. Would be nice to store these separately.

File: caffe2/proto/torch.proto [hunk 3]
```diff
@@ -416,11 +466,36 @@ message TypeProto {
...
   }
 
+  // Sequence type: List, Tuple
+  message Sequence {
+    // Type for each element.
```
Comment (li-roy) [inline caffe2/proto/torch.proto]: I think we need two different cases for list and tuple; all elements in a list are of the same type, but in a tuple, we each element can be of a different type.

File: caffe2/proto/torch.proto [hunk 4]
```diff
@@ -383,6 +421,17 @@ message TensorProto {
...
   // UINT32 or UINT64
   repeated uint64 uint64_data = 11 [packed = true];
+
+  // External data by file name
+  optional string external_data = 13;
+
+  // If two tensors represent the same weights/content, use alias.
+  // Must exist a TensorProto named alias in the initializer list.
+  // To avoid the duplicate tensor in attribute, such as value in Constant node.
+  optional string alias = 16;
```
Comment (li-roy) [inline caffe2/proto/torch.proto]: Is this necessary? If external_data is the same for two tensors, then we know share the same storage.

File: caffe2/proto/torch.proto [hunk 5]
```diff
@@ -383,6 +421,17 @@ message TensorProto {
   // When this field is present, the data_type field MUST be
   // UINT32 or UINT64
   repeated uint64 uint64_data = 11 [packed = true];
+
```
Comment (li-roy) [inline caffe2/proto/torch.proto:427]: There's two additional things that we store for Parameters that don't have a place to go: is_buffer and requires_grad. Both are booleans.

File: caffe2/proto/torch.proto [hunk 6]
```diff
@@ -174,6 +177,9 @@ message NodeProto {
 
   // A human-readable documentation for this node. Markdown is allowed.
   optional string doc_string = 6;
+
```
Comment (li-roy) [inline caffe2/proto/torch.proto:180]: Peter is making a change to store the schema (just a string) in the doc_string for now. We could add an additional scheme field here for that.

Comment (li-roy) [inline caffe2/proto/torch.proto:180]: We could use an additional string field as well, to store the schema of a script module.

File: caffe2/proto/torch.proto (hunk 3 above)
Comment (houseroad) [reply caffe2/proto/torch.proto]: I see. Let me modify this part.

File: caffe2/proto/torch.proto (hunk 2 above)
Comment (houseroad) [reply caffe2/proto/torch.proto:479]: Good point, I will also add `stride` field into TensorShapeProto.

File: caffe2/proto/torch.proto (hunk 6 above)
Comment (houseroad) [reply caffe2/proto/torch.proto:180]: Actually, we can use op_type to store the schema, and store the node type as one annotation, with name "IR_NODE_TYPE"

File: caffe2/proto/torch.proto (hunk 4 above)
Comment (houseroad) [reply caffe2/proto/torch.proto]: This provide another indirection if we want to save everything in one file :-)

File: caffe2/proto/torch.proto (hunk 1 above)
Comment (houseroad) [reply caffe2/proto/torch.proto]: Good catch

File: caffe2/proto/torch.proto [hunk 7]
```diff
@@ -401,7 +461,10 @@ message TensorShapeProto {
     // for pre-defined dimension denotations.
     optional string denotation = 3;
   };
+  // To represent a scalar, using one element dim, and dim[0].dim_value = 0
```
Comment (li-roy) [inline caffe2/proto/torch.proto:464]: Wouldn't this be the same as a one dimensional tensor with size (0,)?

File: caffe2/proto/torch.proto [hunk 8]
```diff
@@ -416,11 +479,39 @@ message TypeProto {
     optional TensorShapeProto shape = 2;
   }
 
+  // Sequence type: List, Tuple
+  message Sequence {
```
Comment (li-roy) [inline caffe2/proto/torch.proto:483]: I think it would be better to have separate messages for List and Tuple and get rid of Sequence because we know Lists have one elem_type and Tuple have multiple. Is there a specific reason we need to support undefined sequence type?

File: caffe2/proto/torch.proto [hunk 9]
```diff
@@ -383,6 +424,25 @@ message TensorProto {
...
+
+  // Additional annotations.
+  repeated AttributeProto annotation = 17;
+
+  // Device info
+  optional DeviceOption device_detail = 51;
+
+  // For PyTorch serialized tensor.
+  optional int64 require_gradient = 52;
+  optional int64 is_bool = 53;
```
Comment (li-roy) [inline caffe2/proto/torch.proto:445]: /is_bool/is_buffer/

**Review by li-roy — COMMENTED** [review]
Looks good, I think this will satisfy everything. Left a some minor comments/questions.

File: caffe2/proto/torch.proto (hunk 8 above)
Comment (houseroad) [reply caffe2/proto/torch.proto:483]: Yeah, because we try to make it onnx compatible. In onnx(-ml) it only supports sequence with the same data type. Let's make it backward compatible. :-)

File: caffe2/proto/torch.proto (hunk 9 above)
Comment (houseroad) [reply caffe2/proto/torch.proto:445]: oops ><

File: caffe2/proto/torch.proto (hunk 7 above)
Comment (houseroad) [reply caffe2/proto/torch.proto:464]: Good point, let's make sure, if it is scalar, `dim` is absent.
